File: Python/refer-offer/cong_shang_dao_xia_da_yin_er_cha_shu_iii_lcof.py
from typing import *

class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

def levelOrder(root: TreeNode) -> List[List[int]]:
    if not root: return []
    res, queue, flag = list(), Deque(), 0
    queue.append(root)
    while queue:
        temp = list()
        for _ in range(len(queue), 0, -1):
            node = queue.popleft()
            if node.left: queue.append(node.left)
            if node.right: queue.append(node.right)
            temp.append(node.val)
        
        if flag % 2 == 1:
            temp.reverse()
        res.append(temp)
        flag += 1
    
    return res

File: Java/src/com/dbc/algorithm/ReferOffer/CongShangDaoXiaDaYinErChaShuIIILcof.java
package com.dbc.algorithm.ReferOffer;

import java.util.*;

public class CongShangDaoXiaDaYinErChaShuIIILcof {
    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode(int x) {
            val = x;
        }
    }

    public List<List<Integer>> levelOrder(TreeNode root) {
        if (root == null) return new ArrayList<>();
        Queue<TreeNode> queue = new LinkedList<>();
        List<List<Integer>> res = new ArrayList<>();
        queue.add(root);
        int flag = 0;
        while (!queue.isEmpty()) {
            List<Integer> temp = new ArrayList<>();
            for (int i = queue.size(); i > 0; i--) {
                TreeNode node = queue.poll();
                if (node.left != null) queue.offer(node.left);
                if (node.right != null) queue.offer(node.right);
                temp.add(node.val);
            }
            if (flag % 2 == 1) {
                Collections.reverse(temp);
            }
            res.add(temp);
            flag++;
        }
        return res;
    }
}
